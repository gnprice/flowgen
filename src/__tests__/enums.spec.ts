import { compiler, beautify } from "..";
import "../test-matchers";

it("should handle empty enums", () => {
  const ts = `enum Empty { }`;
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(beautify(result)).toMatchSnapshot("class");
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should handle basic enums", () => {
  const ts = `enum Label {
    LABEL_OPTIONAL,
    LABEL_REQUIRED,
    LABEL_REPEATED,
  }
type A = Label
type B = Label.LABEL_OPTIONAL

// Including enums in namespaces.
namespace n { export enum E { EM } }
type AA = n.E;
type BB = n.E.EM;

// TODO: This breaks; the first declaration gets lost.  Should merge instead.
// // Including enums with multiple declarations.
// enum ED { A, B }
// // enum ED { C = B + 1, D } // TODO (but not in .d.ts): initializer expression
// enum ED { C = 2, D }
// type EDT = ED.A | ED.C | ED.D;

// TODO: This breaks because we generate separate, conflicting declarations of EN.
// // Including enums that are also namespaces.
// namespace EN { export const nsmem = 1; export type T = string; }
// enum EN { emem = 2 }
// type X = EN;
// type Y = typeof EN.nsmem;
// type Z = EN.emem;
// type W = EN.T;

`;
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(beautify(result)).toMatchSnapshot("class");
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should handle number enums", () => {
  const ts = `enum Label {
    ONE = 1,
    TWO = 2,
    THREE = 3,
    NEGATIVE = -123,
    DECIMAL = 3.14,
  }
type A = Label
type B = Label.TWO
`;
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(beautify(result)).toMatchSnapshot("class");
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should handle string enums", () => {
  const ts = `enum Label {
    LABEL_OPTIONAL = 'LABEL_OPTIONAL',
    LABEL_REQUIRED = 'LABEL_REQUIRED',
    LABEL_REPEATED = 'LABEL_REPEATED',
  }
type A = Label
type B = Label.LABEL_REQUIRED
`;
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(beautify(result)).toMatchSnapshot("class");
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should handle importing enum types", () => {
  const results = compiler.compileDefinitionFiles(
    [
      "src/__tests__/snippet/export-enum-file.ts",
      "src/__tests__/snippet/import-enum-type-file.ts",
    ],
    {
      quiet: false,
    },
  );
  for (const result of results) {
    expect(beautify(result[1])).toMatchSnapshot("class");
    // TODO: this function only runs flow on one file at a time, so it errors when trying to import
    // expect(result[1]).toBeValidFlowTypeDeclarations();
  }
});

it("should handle importing enums", () => {
  const results = compiler.compileDefinitionFiles(
    [
      "src/__tests__/snippet/export-enum-file.ts",
      "src/__tests__/snippet/reexport-enum-file.ts",
      "src/__tests__/snippet/import-enum-file.ts",
    ],
    {
      quiet: false,
    },
  );
  for (const result of results) {
    expect(beautify(result[1])).toMatchSnapshot("class");
    // TODO: this function only runs flow on one file at a time, so it errors when trying to import
    // expect(result[1]).toBeValidFlowTypeDeclarations();
  }
});

it("should handle enum on imported module", () => {
  const results = compiler.compileDefinitionFiles(
    [
      "src/__tests__/snippet/export-enum-file.ts",
      "src/__tests__/snippet/import-enum-module-file.ts",
    ],
    {
      quiet: false,
    },
  );
  for (const result of results) {
    expect(beautify(result[1])).toMatchSnapshot("class");
    // TODO: this function only runs flow on one file at a time, so it errors when trying to import
    // expect(result[1]).toBeValidFlowTypeDeclarations();
  }
});
