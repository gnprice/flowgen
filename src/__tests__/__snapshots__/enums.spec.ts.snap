// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle basic enums: class 1`] = `
"declare var Label: {|
  +LABEL_OPTIONAL: 0, // 0
  +LABEL_REQUIRED: 1, // 1
  +LABEL_REPEATED: 2, // 2
|};
declare type A = $Values<typeof Label>;
declare type B = typeof Label.LABEL_OPTIONAL;
declare var n: typeof npm$namespace$n;

declare var npm$namespace$n: {|
  E: typeof n$E,
|};

declare export var n$E: {|
  +EM: 0, // 0
|};
declare type AA = $Values<typeof n$E>;
declare type BB = typeof n$E.EM;
"
`;

exports[`should handle empty enums: class 1`] = `
"declare var Empty: {||};
"
`;

exports[`should handle enum on imported module: class 1`] = `
"declare export var Label: {|
  +A: \\"A\\", // \\"A\\"
  +B: \\"B\\", // \\"B\\"
|};
export type LabelA = typeof Label.A;
export type LLabel = $Values<typeof Label>;
"
`;

exports[`should handle enum on imported module: class 2`] = `
"import * as m from \\"./export-enum-file\\";
declare export function foo(label: $Values<typeof m.Label>): void;
"
`;

exports[`should handle importing enum types: class 1`] = `
"declare export var Label: {|
  +A: \\"A\\", // \\"A\\"
  +B: \\"B\\", // \\"B\\"
|};
export type LabelA = typeof Label.A;
export type LLabel = $Values<typeof Label>;
"
`;

exports[`should handle importing enum types: class 2`] = `
"import typeof { Label } from \\"./export-enum-file\\";
declare export function foo(label: $Values<Label>): void;
"
`;

exports[`should handle importing enums: class 1`] = `
"declare export var Label: {|
  +A: \\"A\\", // \\"A\\"
  +B: \\"B\\", // \\"B\\"
|};
export type LabelA = typeof Label.A;
export type LLabel = $Values<typeof Label>;
"
`;

exports[`should handle importing enums: class 2`] = `
"declare export * from \\"./export-enum-file\\";
"
`;

exports[`should handle importing enums: class 3`] = `
"import { Label as ImportedLabel, LabelA, LLabel } from \\"./export-enum-file\\";
import {
  Label as ImportedLabelR,
  LabelA as LabelAR,
  LLabel as LLabelR,
} from \\"./reexport-enum-file\\";
declare export function foo(label: $Values<typeof Label>): void;
export type WM = typeof Label.A;
export type X = typeof Label.A;
export type Y = $Values<typeof Label>;
export type WR = $Values<typeof Label>;
export type WRM = typeof Label.A;
export type XR = typeof Label.A;
export type YR = $Values<typeof Label>;
"
`;

exports[`should handle number enums: class 1`] = `
"declare var Label: {|
  +ONE: 1, // 1
  +TWO: 2, // 2
  +THREE: 3, // 3
  +NEGATIVE: -123, // -123
  +DECIMAL: 3.14, // 3.14
|};
declare type A = $Values<typeof Label>;
declare type B = typeof Label.TWO;
"
`;

exports[`should handle string enums: class 1`] = `
"declare var Label: {|
  +LABEL_OPTIONAL: \\"LABEL_OPTIONAL\\", // \\"LABEL_OPTIONAL\\"
  +LABEL_REQUIRED: \\"LABEL_REQUIRED\\", // \\"LABEL_REQUIRED\\"
  +LABEL_REPEATED: \\"LABEL_REPEATED\\", // \\"LABEL_REPEATED\\"
|};
declare type A = $Values<typeof Label>;
declare type B = typeof Label.LABEL_REQUIRED;
"
`;
